sw/device/lib/dif/dif_rstmgr.c

```diff
@@ -0,0 +1,215 @@
+                   RSTMGR_USB_REGEN_USB_REGEN,
+               "SPI and USB regen bit offsets are not the same");
+_Static_assert(RSTMGR_RST_SPI_DEVICE_N_RST_SPI_DEVICE_N ==
+                   RSTMGR_RST_USB_N_RST_USB_N,
+               "SPI and USB reset bit offsets are not the same");
+
+#define RSTMGR_PERIPHERAL_REGEN RSTMGR_SPI_DEVICE_REGEN_SPI_DEVICE_REGEN
+#define RSTMGR_PERIPHERAL_RST_N RSTMGR_RST_SPI_DEVICE_N_RST_SPI_DEVICE_N
+
+static const bitfield_field32_t kRstmgrCausesField = {
+    .mask = RSTMGR_RESET_INFO_CAUSE_MASK,
+    .index = RSTMGR_RESET_INFO_CAUSE_OFFSET,
+};
+
+// Regen registers are "write 1 to clear" = write 1 to lock.
+static const ptrdiff_t kRstmgrRegenOffsetsMap[kDifRstmgrPeripheralLast + 1] = {
+        [kDifRstmgrPeripheralSpi] = RSTMGR_SPI_DEVICE_REGEN_REG_OFFSET,
+        [kDifRstmgrPeripheralUsb] = RSTMGR_USB_REGEN_REG_OFFSET,
+};
+
+// Reset registers, when set to 0, peripheral is held in reset.
+static const ptrdiff_t kRstmgrResetOffsetsMap[kDifRstmgrPeripheralLast + 1] = {
+        [kDifRstmgrPeripheralSpi] = RSTMGR_RST_SPI_DEVICE_N_REG_OFFSET,
+        [kDifRstmgrPeripheralUsb] = RSTMGR_RST_USB_N_REG_OFFSET,
+};
+
+static void rstmgr_software_reset_hold(mmio_region_t base_addr,
+                                       ptrdiff_t reg_offset) {
+  uint32_t bitfield = bitfield_bit32_write(0, RSTMGR_PERIPHERAL_RST_N, false);
+  mmio_region_write32(base_addr, reg_offset, bitfield);
+}
+
+static bool rstmgr_software_reset_is_held(mmio_region_t base_addr,
+                                          ptrdiff_t reg_offset) {
+  uint32_t bitfield = mmio_region_read32(base_addr, reg_offset);
+  return !bitfield_bit32_read(bitfield, RSTMGR_PERIPHERAL_RST_N);
+}
+
+static void rstmgr_software_reset_release(mmio_region_t base_addr,
+                                          ptrdiff_t reg_offset) {
+  uint32_t bitfield = bitfield_bit32_write(0, RSTMGR_PERIPHERAL_RST_N, true);
+  mmio_region_write32(base_addr, reg_offset, bitfield);
+}
+
+static void rstmgr_software_reset_lock(mmio_region_t base_addr,
+                                       ptrdiff_t reg_offset) {
+  uint32_t bitfield = bitfield_bit32_write(0, RSTMGR_PERIPHERAL_REGEN, true);
+  mmio_region_write32(base_addr, reg_offset, bitfield);
+}
+
+static bool rstmgr_software_reset_locked(mmio_region_t base_addr,
+                                         ptrdiff_t reg_offset) {
+  uint32_t bitfield = mmio_region_read32(base_addr, reg_offset);
+  return !bitfield_bit32_read(bitfield, RSTMGR_PERIPHERAL_REGEN);
+}
+
+dif_rstmgr_result_t dif_rstmgr_init(dif_rstmgr_param_t params,
+                                    dif_rstmgr_t *handle) {
+  if (handle == NULL) {
+    return kDifRstmgrBadArg;
+  }
+
+  handle->params = params;
+
+  return kDifRstmgrOk;
+}
+
+dif_rstmgr_result_t dif_rstmgr_reset(const dif_rstmgr_t *handle) {
+  if (handle == NULL) {
+    return kDifRstmgrBadArg;
+  }
+
+  mmio_region_t base_addr = handle->params.base_addr;
```

Comment:
Yup, I understand, I just thought that because @mcy  change (DIF template) has landed - everyone has agreed on it, so I simply copy/pasted it (I admit, I haven't followed the that PR closely).

Reply:
I think we are going forward with Miguel's template, so I will mark this comment as addressed (keeping the params).